Item overlay customization API

Leo40Git opened the pull request:
A more specialized approach to customizing item overlays, with the generic approach I implemented in #881 still intact.

### Features
- Configuring every component of the item overlay
  - Count label (`ItemLabelInfo`) - visibility, contents, color
  - Damage bar (`ItemDamageBarInfo`) - visibility, fill factor, color
  - Cooldown overlay (`ItemCooldownInfo`) - visibility, fill factor, color
- Rendering before item overlay *or* replacing it entirely (`ItemOverlayRenderer.Pre`)
  - Identical to #881's `ItemOverlayRenderer`
- Appending to the item overlay (`ItemOverlayRenderer.Post`)

### To-do
**All done!** Ready for merging!!!
- [x] Test mod
  - [x] Count label
  - [x] Damage bar
  - [x] Cooldown overlay
  - [x] Item overlay replacement
  - [x] Item overlay appending
- [x] Javadocs for the 3 component property interfaces

Leo40Git committed:
Item overlay customization API

Leo40Git committed:
Fix CustomItemOverlayRenderer's code example

Leo40Git committed:
CustomItemOverlayRenderer: allow configurable cancelling

Leo40Git commented:
Note that I'm a bit iffy about the first draft here, since it requires an `@Overwrite` mixin to work. I *may* be able to remove the `@Overwrite` if I remove the "multiple durability bars" feature and split it into... a *lot* of `@ModifyVariable`s and `@Redirect`s.

Leo40Git committed:
damn you, checkstyle

Leo40Git committed:
Split default implementations into separate classes

Leo40Git committed:
SingleDurabilityBarProperties: optimization??

Leo40Git committed:
Minor refactors

Move default implementation classes to "impl"
Move default implementation instance fields into interfaces (DefaultXXX.INSTANCE -> XXX.DEFAULT)

liach commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/CooldownOverlayProperties.java:
```diff
@@ -0,0 +1,29 @@
+	CooldownOverlayProperties DEFAULT = new DefaultCooldownOverlayProperties();
+
+	boolean isVisible(ItemStack stack);
+	float getFillFactor(ItemStack stack);
```
maybe `getProgress`/`getFillProgress`?

shartte replied:
Since this is purely graphical and the overlay can be used to represent anything (not just cooldowns), I think fill factor is fine. It should just say how much of the item is filled with the overlay.

liach commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItemRenderer.java:
```diff
@@ -0,0 +1,125 @@
+	 * @author ADudeCalledLeo
+	 */
+	@SuppressWarnings("deprecation")
+	@Overwrite
```
this piece of mixin seems vulnerable to minecraft updates due to its overwrite nature, though i cannot figure out a way to improve it unfortunately.

liach reviewed (commented):
Great on first sight. I probably should browse again to make sure there is no environment issues.

Leo40Git committed:
Un-invert DurabilityBarProperties.getCount

Leo40Git committed:
Use full variable names

Leo40Git committed:
Expand decompiled one-letter params to full names

shartte commented:
@Leo40Git I made a PR to your PR here: https://github.com/Leo40Git/fabric/pull/1
Should I rebase that on your latest changes or do you want to have a look first?
- It moves the replicated rendering code out to a Helper class and makes that accessible to modders piecemeal.
- Changes the Mixin to an @Inject
- Implements a testmod for custom durability bars and custom cooldown overlays

Leo40Git commented:
Rebase it, I'm totally taking the Javadocs and test mod. 
I'm not sure if the helper class would be an improvement here yet, I'll wait for the reviewers' thoughts about it.

shartte committed:
Added a testmod for custom durability bars and cooldown overlays.

shartte commented:
@Leo40Git I rebased it and split off any refactoring of the Mixin into a separate branch (will PR separately). The PR now only contains the testmod. Testmod also still needs a test for custom label and fully custom renderer.

Leo40Git committed:
Fix Javadoc of CustomItemOverlayRenderer.renderOverlay

Leo40Git committed:
Only allocate MatrixStack once

Leo40Git committed:
Add source comments for default impls

Leo40Git committed:
Fix test mod durability bars being inverted

Leo40Git committed:
Fix testmod and default durability bars being inverted

Leo40Git committed:
consistency

Leo40Git committed:
Fix default count label and durability bar properties

Testmod: Fix Hidden Cooldown item not having localization
Testmod: Add count label test: Obfuscated Count Label item

Leo40Git committed:
Split CustomItemOverlayRenderer into "before" and "after" stages

PreItemOverlayRenderer is equivalent to CustomItemOverlayRenderer
PostItemOverlayRenderer is identical, except it's called *after* the overlay is rendered and (thus, obviously) can't cancel it

Leo40Git committed:
Split default implementations for pre- and post-item overlay renderers into separate classes

Move item overlay related impl classes to their own sub-package

Leo40Git commented:
maybe i should move this stuff to the `item-api` module

liach commented:
This is really rendering api, not item api

Leo40Git commented:
Now that #956 is merged, should I change this PR to use `FabricItemSettings` rather than storing stuff on the `Item` directly?
I'll probably close this PR and make a new one because *holy shit is GitHub struggling*.

Leo40Git commented:
Other changes for a theoretical v3 are dropping multiple durability bars and changing my `@Overwrite` mixin into a more agreeable mishmash of `@Inject`s and `@ModifyArg`s.

i509VCB commented:
If you could provide a concept of your mishmash of injectors I would look at that.

Leo40Git committed:
Rewrite mixin to yeet @Overwrite

TODO: Refactor DurabilityBarProperties to remove multi-bar capabilities (RIP)

Leo40Git commented:
@i509VCB One `@Overwrite`-less mixin, fresh off the presses!

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/DefaultCooldownOverlayProperties.java:
```diff
@@ -0,0 +1,47 @@
+import net.fabricmc.fabric.api.client.rendereregistry.v1.item.CooldownOverlayProperties;
+
+public class DefaultCooldownOverlayProperties implements CooldownOverlayProperties {
+	protected float getCooldownAmount(ItemStack stack) {
```
Possibly pass minecraft client here. Yes you can get instance (shedaniel) but if mojang changes their mind in the future then this doesn't break.

Leo40Git replied:
99e3c6d938cd4006e446364364a24617d50a9c99

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItemRenderer.java:
```diff
@@ -0,0 +1,151 @@
+
+@Mixin(ItemRenderer.class)
+public abstract class MixinItemRenderer {
+	// This design means multiple durability bars are impossible. Too bad!
```
Rofl reminds me of HL2 comments in leaked source code.

Leo40Git replied:
this was inspired by the "Too bad!" comments found in the TF2 source code leak, so you're not far off

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItemRenderer.java:
```diff
@@ -0,0 +1,151 @@
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At("HEAD"), cancellable = true)
+	public void preOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel, CallbackInfo ci) {
```
These all look good, would be nice if you have some docs explaining the purpose behind some of these injects and redirects.

Leo40Git replied:
done kinda 9c8f073c72433a542df6c8558d3f4eae2157b754

i509VCB commented on fabric-renderer-registries-v1/src/testmod/java/net/fabricmc/fabric/test/renderregistry/client/countlabel/ObfuscatedCountLabelProperties.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	@Override
+	public String getContents(ItemStack stack, String override) {
```
Based on how this looks, would it be worthwhile to possibly return text in this method?

Leo40Git replied:
I highly doubt someone's gonna want translatable count labels, but I could do it
edit: oh right, style/formatting stuff

Leo40Git replied:
16 hours later... I still don't see the point of making it return a `Text`, since the code would need to convert it back into a `String` for rendering purposes anyway.

i509VCB replied:
Doesn't the text renderer convert it to OrderedText at some point. And there is a text renderers method specific to text which does that conversion.

Leo40Git committed:
add descriptive comments to mixin

Leo40Git committed:
RIP multiple durability bars

also the water/lava bucket
also also disco ball used to use an AWT method, that's been yeeted as well

Leo40Git committed:
Add MinecraftClient parameter to CooldownOverlayProperties methods

Leo40Git committed:
Change CountLabelProperties.getContents return type to Text

remove note about "gross redirect" since now it *has* to be a redirect

Leo40Git committed:
"count label label", yes good job me

Leo40Git committed:
split long line into vars

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+	PreItemOverlayRenderer fabric_getPreItemOverlayRenderer();
+	void fabric_setPreOverlayRenderer(PreItemOverlayRenderer pior);
+
+	PostItemOverlayRenderer fabric_getPostItemOveralyRenderer();
```
```suggestion
	PostItemOverlayRenderer fabric_getPostItemOverlayRenderer();
```

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+
+public interface ItemOverlayExtensions {
+	CountLabelProperties fabric_getCountLabelProperties();
+	void fabric_setCountLabelProperties(CountLabelProperties clp);
```
```suggestion
	void fabric_setCountLabelProperties(CountLabelProperties properties);
```

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+	void fabric_setCountLabelProperties(CountLabelProperties clp);
+
+	DurabilityBarProperties fabric_getDurabilityBarProperties();
+	void fabric_setDurabilityBarProperties(DurabilityBarProperties dbp);
```
```suggestion
	void fabric_setDurabilityBarProperties(DurabilityBarProperties properties);
```

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+	void fabric_setDurabilityBarProperties(DurabilityBarProperties dbp);
+
+	CooldownOverlayProperties fabric_getCooldownOverlayProperties();
+	void fabric_setCooldownOverlayProperties(CooldownOverlayProperties cop);
```
```suggestion
	void fabric_setCooldownOverlayProperties(CooldownOverlayProperties properties);
```

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+	void fabric_setCooldownOverlayProperties(CooldownOverlayProperties cop);
+
+	PreItemOverlayRenderer fabric_getPreItemOverlayRenderer();
+	void fabric_setPreOverlayRenderer(PreItemOverlayRenderer pior);
```
```suggestion
	void fabric_setPreOverlayRenderer(PreItemOverlayRenderer renderer);
```

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/item/ItemOverlayExtensions.java:
```diff
@@ -0,0 +1,40 @@
+	void fabric_setPreOverlayRenderer(PreItemOverlayRenderer pior);
+
+	PostItemOverlayRenderer fabric_getPostItemOveralyRenderer();
+	void fabric_setPostOverlayRenderer(PostItemOverlayRenderer pior);
```
```suggestion
	void fabric_setPostOverlayRenderer(PostItemOverlayRenderer renderer);
```

Leo40Git committed:
haykam refactors

Leo40Git committed:
Javadoc for component properties interfaces

Update old example code snippet in PreItemOverlayRenderer javadoc

Leo40Git commented:
Looking into the behavior quirk mentioned in `CountLabelProperties.isVisible`.
In the meantime, I'll write some testmods

Leo40Git committed:
(Broken) testmod stuff for pre/post-rendering

For some reason DrawableHelper and Tessellator are refusing to help me in my quest to draw 4 lines onto the screen.

Leo40Git commented:
assistance required, because apparently editing a method parameter or *rendering a square* just isn't working for me

haykam821 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/CooldownOverlayProperties.java:
```diff
@@ -0,0 +1,53 @@
+	 * @param client current {@link MinecraftClient} instance
+	 * @return overlay fill factor, between 0 and 1 (inclusive)
+	 */
+	float getFillFactor(ItemStack stack, MinecraftClient client);
```
```suggestion
	float getBarPercent(ItemStack stack, MinecraftClient client);
```

Leo40Git replied:
See https://github.com/FabricMC/fabric/pull/916#discussion_r460462920

Leo40Git replied:
Also, `getBarPercent` would fit `DurabilityBarProperties` more... but I'd rather keep it consistent

shartte replied:
A value between 0 and 1 is not a percentage.

Leo40Git committed:
Hack around count label always being rendered if contents overriden

Thanks to @.Gimpansor for helping with the initial "countVisibleCondHack" thing

Leo40Git committed:
comment shtuff

Leo40Git committed:
get pre/post-renderer testmods working

APPARENTLY "Tessellator.getInstance().draw()" is out, the hot new thing is "BufferRenderer.render(BufferBuilder)"

Leo40Git commented:
IT IS DONE
Should I try resolving the conflicts or...?

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,65 @@
+public final class ItemOverlayRendererRegistry {
+	private ItemOverlayRendererRegistry() { }
+
+	public static CountLabelProperties getCountLabelProperties(Item item) {
```
All of these should be ItemConvertible instead of Item.

This is so people can pipe a block into the method.

Leo40Git replied:
time to regex this shit up

i509VCB replied:
Rofl in the method just call `asItem` and null check it lol

Leo40Git replied:
oh oops i forgot the null check part

Leo40Git replied:
so in getters, if `ItemConvertible.asItem()` returns `null`, should i also return `null` or throw an `IllegalArgumentException`

i509VCB replied:
I usually just Objects.requireNonNull the item convertible instance and then the result of `asItem`

Leo40Git replied:
1acbb44 :)

Leo40Git committed:
ItemOverlayRendererRegistry: add ItemConvertible overloads

Leo40Git committed:
ItemOverlayRendererRegistry: null chks

Leo40Git committed:
NO nulls are allowed in my CHRISTIAN REGISTRY

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/CountLabelProperties.java:
```diff
@@ -0,0 +1,53 @@
+ * This interface allows you to modify the count label that is displayed on item stacks in inventories.
+ */
+public interface CountLabelProperties {
+	CountLabelProperties DEFAULT = new DefaultCountLabelProperties();
```
Whats the use of this being in the api? Could it just be in DefaultCountLabelProperties?

Leo40Git replied:
honestly I don't even remember why I put this here
I think i thought something along the lines of "mod devs shouldn't touch API implementations"
...and then I said "the default impls are for mixing into", so good job me

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/CountLabelProperties.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * This interface allows you to modify the count label that is displayed on item stacks in inventories.
+ */
+public interface CountLabelProperties {
```
`ItemLabe`l? Count doesnt really say what it does. Also not sure if `Properties` is quite right

Leo40Git replied:
"count label" is what the Yarn mappings call the parameter that overrides the label's contents in `ItemRenderer.renderGuiItemIcon`, FYI
but "item label" is better, really

Leo40Git replied:
it's now `ItemLabelInfo`

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/CooldownOverlayProperties.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * This interface allows you to modify the cooldown overlay that is displayed on item stacks in inventories.
+ */
+public interface CooldownOverlayProperties {
```
Again, here `ItemCooldownOverlay` might be better.

Leo40Git replied:
`ItemCooldownInfo`

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/DurabilityBarProperties.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * This interface allows you to modify the durability bar that is displayed on item stacks in inventories.
+ */
+public interface DurabilityBarProperties {
```
Again, possibly add Item prefix.

Leo40Git replied:
`ItemDamageBarProperties`

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/PostItemOverlayRenderer.java:
```diff
@@ -0,0 +1,37 @@
+ * and cooldown overlays).
+ */
+@FunctionalInterface
+public interface PostItemOverlayRenderer {
```
What about making just 1 outer class with 2 inner classes? ItemOverlayRenderer.Pre/Post?

Leo40Git replied:
so should the outer class be a `final class` or an `interface`?

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/item/PreItemOverlayRenderer.java:
```diff
@@ -0,0 +1,52 @@
+	 * enchanted glint is rendered.
+	 * @return {@code true} to cancel Vanilla's overlay rendering.
+	 */
+	@Environment(EnvType.CLIENT)
```
Make the whole class Client only.

Leo40Git replied:
I honestly don't remember when I put that annotation there
Probably a holdover from #881's `CustomItemOverlayRenderer`

Leo40Git replied:
https://github.com/FabricMC/fabric/pull/881#discussion_r459649403 #blamei5

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItem.java:
```diff
@@ -0,0 +1,88 @@
+
+@Mixin(Item.class)
+public abstract class MixinItem implements ItemOverlayExtensions {
+	@Unique private CountLabelProperties countLabelProperties = CountLabelProperties.DEFAULT;
+	@Unique private DurabilityBarProperties durabilityBarProperties = DurabilityBarProperties.DEFAULT;
+	@Unique private CooldownOverlayProperties cooldownOverlayProperties = CooldownOverlayProperties.DEFAULT;
+	@Unique private PreItemOverlayRenderer preItemOverlayRenderer = PreItemOverlayRenderer.DEFAULT;
+	@Unique private PostItemOverlayRenderer postItemOverlayRenderer = PostItemOverlayRenderer.DEFAULT;
```
Would null be better than the default? would remove quite a lot of bascily empty classes

Leo40Git replied:
yeeting vanillacopy and empty classes is next on the agenda

Leo40Git replied:
RIP in hecc default implementation classes, 2020-2020  
their last words were "why did you even make us in the first place"

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItem.java:
```diff
@@ -0,0 +1,88 @@
+import net.fabricmc.fabric.impl.client.renderer.registry.item.ItemOverlayExtensions;
+
+@Mixin(Item.class)
+public abstract class MixinItem implements ItemOverlayExtensions {
```
Not a bit fan of doing this, mojang are moving client stuff out of Block/Item (Think render layers). And I think client/common code should have clear boundaries.

Doing this many hashmap lookups might become a peformace issue?

Leo40Git replied:
it's looking up a `HashMap` *every frame* an item stack is in your hotbar/open screen and caching won't really help here since what's the chance *all stacks on the screen* are of the same item
and item rendering (for 2D textures) is notoriously unefficent already (just ask jelly)

Leo40Git replied:
but I've never actually done my own benchmarking, so
I'll just shift this over to use a few `Object2ObjectOpenHashMap`s on the API side

Leo40Git replied:
https://github.com/FabricMC/fabric/pull/881#discussion_r452966900 apparently you should blame i5 for this "store renderer stuff directly on the `Item` instance" design

Leo40Git replied:
Moved to `IdentityHashMap`s, which compare equality via `==` instead of `equals`. *Should* have little-to-no impact on performance.

modmuss50 reviewed (changes requested):
Make sure to put client only on basiclly all of this, IMO the client/server split needs to be very clear.

Not a big fan of the class names for the iterfaces, we can disucss better names.

Id use null over the default impl's, removes a lot of fluff.

Leo40Git commented:
> Make sure to put client only on basiclly all of this, IMO the client/server split needs to be very clear.

will do tomorrow

> Not a big fan of the class names for the iterfaces, we can disucss better names.

is it bikeshed hours :yeefuckinhaw: /s

> Id use null over the default impl's, removes a lot of fluff.

I did the default impl stuff to allow other mods to change default behavior via mixins without possibly breaking everything

modmuss50 commented:
> allow other mods to change default behavior via mixins

If this is in scope (im not sure it is) it should ideally be provided via the api. The long term idea is to make it so most mods dont need to use mixin. Now is the best time to design the api if this is something that is wanted.

Leo40Git commented:
hm. possibly have a "global"/fallback implementation for every interface, if both individual and fallback impls are null default to vanilla behavior?

Leo40Git commented:
Here's the new names for the interfaces:
* `CountLabelProperties` -> `ItemLabelInfo`
* `DurabilityBarProperties` -> `ItemDamageBarInfo` (changed to be consistent with `ItemStack.damage`)
* `CooldownOverlayProperties` -> `ItemCooldownOverlayInfo`
* `PreItemOverlayRenderer` -> `ItemOverlayRenderer.Pre`
* `PostItemOverlayRenderer` -> `ItemOverlayRenderer.Post`

Leo40Git committed:
Yeet default implementations

Add Environment(CLIENT) annotations

Leo40Git committed:
Rename interfaces

Leo40Git committed:
be consistent, damnit

Leo40Git committed:
Moved overlay storage from Item instances

The new map-based storage uses IdentityHashMaps, which *theoretically* should result in little-to-no performance decrease.
ItemOverlayRendererRegistry no longer has getter methods (which only the implementation itself used)
Also it actually compiles now oops

Leo40Git committed:
Annotate the other 3 classes here with Environment(CLIENT)

Leo40Git commented:
We've entered the endgame, bois  
And by that I meant bikeshedding about class names /s

But no seriously, I'm pretty much done.  
Could you do the discussions in the Fabricord's #api? This PR is apparently way too big for my poor laptop because fuck is it lagging.

Only real "functional" things I can do now are multiple damage bars (which I'm still iffy on, since it *can* be done without a total `@Overwrite` but I *do* need to duplicate the damage bar code at least once) and resolving the merge conflicts.

edit: Multiple damage bars are definitely not coming to this PR (might make a separate mini-library mod for it), and the merge conflicts have been fixed.

Leo40Git committed:
Allow setting default (fallback) item overlay components

Leo40Git committed:
ItemCooldownInfo -> ItemCooldownOverlayInfo

Update references of "durability bar" to "damage bar"

Leo40Git committed:
remove 1-class package

remove comments about multiple damage bars, it's been definitively dropped

Leo40Git committed:
Fix some mistakes in MixinItemRenderer

- pre-rendering allocated another MatrixStack
- bar was visible on damageable items by default (instead of damaged items)

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.16' into 1.16-itemoverlay-v2

# Conflicts:
#	fabric-renderer-registries-v1/build.gradle
#	fabric-renderer-registries-v1/src/testmod/resources/fabric.mod.json

Leo40Git committed:
Resolve merge conflicts

Leo40Git commented:
Okay, let's talk overrides:
- Disallow overriding, throw an exception when an override is attempted
- Allow overriding (with logging), add `has*` methods so mods can check if an item already has an item overlay component.

Leo40Git commented:
closing this, will be available as a separate API at a later date
